src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/Delegation/DelegatedCompletionHelper.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Razor.DocumentMapping;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Protocol.Completion;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.Completion.Delegation;
+
+using SyntaxNode = Microsoft.AspNetCore.Razor.Language.Syntax.SyntaxNode;
+
+/// <summary>
+/// Helper methods for C# and HTML completion ("delegated" completion) that are used both in LSP and cohosting
+/// completion handler code.
+/// </summary>
+internal static class DelegatedCompletionHelper
+{
+    // Ordering should be:
+    // 1. Changes items
+    // 2. Adds items
+    // 3. Filters items
+    private static readonly ImmutableArray<IDelegatedCSharpCompletionResponseRewriter> s_delegatedCSharpCompletionResponseRewriters =
+        [new SnippetResponseRewriter(), new TextEditResponseRewriter(), new DesignTimeHelperResponseRewriter()];
+
+    // Currently we only have one HTML response re-writer. Should we ever need more, we can create a common base and a collection
+    private static readonly HtmlCommitCharacterResponseRewriter s_delegatedHtmlCompletionResponseRewriter = new HtmlCommitCharacterResponseRewriter();
+
+    /// <summary>
+    /// Modifies completion context if needed so that it's acceptable to the delegated language.
+    /// </summary>
+    /// <param name="context">Original completion context passed to the completion handler</param>
+    /// <param name="languageKind">Language of the completion position</param>
+    /// <returns>Possibly modified completion context</returns>
+    /// <remarks>For example, if we invoke C# completion in Razor via @ character, we will not
+    /// want C# to see @ as the trigger character and instead will transform completion context
+    /// into "invoked" and "explicit" rather than "typing", without a trigger character</remarks>
+    public static VSInternalCompletionContext RewriteContext(VSInternalCompletionContext context, RazorLanguageKind languageKind)
+    {
+        Debug.Assert(languageKind != RazorLanguageKind.Razor,
+            $"{nameof(RewriteContext)} should be called for delegated completion only");
+
+        if (context.TriggerKind != CompletionTriggerKind.TriggerCharacter ||
+            context.TriggerCharacter is not { } triggerCharacter)
```

Comment:
Why the `||` on the first line here? Based on the rest of this file, I would have assumed you would put it on the second?

Reply:
Because that's how it was in the original file

https://github.com/dotnet/razor/blob/dab429d0cd6c5b5c7550e5591837b91df4478a6b/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/Delegation/DelegatedCompletionListProvider.cs#L151

I didn't go through every line of code I moved.